wasm/src/proxy_wasm.cc

```diff
@@ -0,0 +1,137 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+#include "proxy_wasm_intrinsics.h"
+#include "context.h"
+#include "json.h"
+#include "malloc.h"
+
+class ExampleRootContext : public RootContext {
+public:
+  explicit ExampleRootContext(uint32_t id, std::string_view root_id) : RootContext(id, root_id) {}
+
+  bool onStart(size_t) override;
+  bool onConfigure(size_t) override;
+  void onTick() override;
+};
+
+class ExampleContext : public Context {
+public:
+  explicit ExampleContext(uint32_t id, RootContext *root) : Context(id, root) {}
+
+  void onCreate() override;
+  FilterHeadersStatus onRequestHeaders(uint32_t headers, bool end_of_stream) override;
+  FilterDataStatus onRequestBody(size_t body_buffer_length, bool end_of_stream) override;
+  void onDone() override;
+  void onLog() override;
+  void onDelete() override;
+
+  opa_eval_ctx_t* eval_ctx_;
+};
+
+// NOTE(sr): we're calling this to ensure that our implementation here is registered
+// with the proxy_wasm_intrinsics.cc globals. The attempt above is suspect.
+// This method is called by the module's Start function (_initialize), which is
+// wired up in the wasm compiler.
+extern "C" OPA_INTERNAL void _register_proxy_wasm(void) {
+  RegisterContextFactory register_ExampleContext(CONTEXT_FACTORY(ExampleContext),
+                                                 ROOT_FACTORY(ExampleRootContext),
+                                                 "opa_root_id");
+}
+
+bool ExampleRootContext::onStart(size_t) {
+  logInfo("onStart");
+  return true;
+}
+
+bool ExampleRootContext::onConfigure(size_t) {
+  logInfo("onConfigure");
+  proxy_set_tick_period_milliseconds(1000); // 1 sec
+  return true;
+}
+
+void ExampleRootContext::onTick() { logTrace("onTick"); }
+
+void ExampleContext::onCreate() {
+  logInfo("onCreate");
+  eval_ctx_ = opa_eval_ctx_new();
+}
+
+FilterHeadersStatus ExampleContext::onRequestHeaders(uint32_t, bool) {
+  logInfo("onRequestHeaders _");
+  auto resp = getRequestHeaderPairs();
+  auto pairs = resp->pairs();
+
+  opa_object_t *hdrs = opa_cast_object(opa_object());
+
+  // TODO(sr): If a header key is repeated, the second one overwrites the first
+  // one. We should figure out if we want to change the input structure, or append
+  // those values separated by ","
+  for (auto& p : pairs) {
+    char *key = (char *) opa_malloc(p.first.size() * sizeof(key));
+    char *val = (char *) opa_malloc(p.second.size() * sizeof(val));
+    memcpy(key, p.first.data(), p.first.size());
+    memcpy(val, p.second.data(), p.second.size());
```

Comment:
https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/84 it took me a while to figure out what was going on there.